moshi/src/main/java/com/squareup/moshi/JsonUtf8Reader.kt

```diff
@@ -0,0 +1,1186 @@
+   */
+  private var valueSource: JsonValueSource? = null
+
+  constructor(source: BufferedSource?) {
+    if (source == null) {
+      throw NullPointerException("source == null")
+    }
+    this.source = source
+    buffer = source.buffer
+    pushScope(JsonScope.EMPTY_DOCUMENT)
+  }
+
+  /** Copy-constructor makes a deep copy for peeking.  */
+  constructor(copyFrom: JsonUtf8Reader) : super(copyFrom) {
+    val sourcePeek = copyFrom.source.peek()
+    source = sourcePeek
+    buffer = sourcePeek.buffer
+    peeked = copyFrom.peeked
+    peekedLong = copyFrom.peekedLong
+    peekedNumberLength = copyFrom.peekedNumberLength
+    peekedString = copyFrom.peekedString
+
+    // Make sure our buffer has as many bytes as the source's buffer. This is necessary because
+    // JsonUtf8Reader assumes any data it has peeked (like the peekedNumberLength) are buffered.
+    try {
+      sourcePeek.require(copyFrom.buffer.size)
+    } catch (e: IOException) {
+      throw AssertionError()
+    }
+  }
+
+  @Throws(IOException::class)
+  override fun beginArray() {
+    var p = peeked
+    if (p == PEEKED_NONE) {
+      p = doPeek()
+    }
+    if (p == PEEKED_BEGIN_ARRAY) {
+      pushScope(JsonScope.EMPTY_ARRAY)
+      pathIndices[stackSize - 1] = 0
+      peeked = PEEKED_NONE
+    } else {
+      throw JsonDataException("Expected BEGIN_ARRAY but was ${peek()} at path $path")
+    }
+  }
+
+  @Throws(IOException::class)
+  override fun endArray() {
+    var p = peeked
+    if (p == PEEKED_NONE) {
+      p = doPeek()
+    }
+    if (p == PEEKED_END_ARRAY) {
+      stackSize--
+      pathIndices[stackSize - 1]++
+      peeked = PEEKED_NONE
+    } else {
+      throw JsonDataException("Expected END_ARRAY but was ${peek()} at path $path")
+    }
+  }
+
+  @Throws(IOException::class)
+  override fun beginObject() {
+    var p = peeked
+    if (p == PEEKED_NONE) {
+      p = doPeek()
+    }
+    if (p == PEEKED_BEGIN_OBJECT) {
+      pushScope(JsonScope.EMPTY_OBJECT)
+      peeked = PEEKED_NONE
+    } else {
+      throw JsonDataException("Expected BEGIN_OBJECT but was ${peek()} at path $path")
+    }
+  }
+
+  @Throws(IOException::class)
+  override fun endObject() {
+    var p = peeked
+    if (p == PEEKED_NONE) {
+      p = doPeek()
+    }
+    if (p == PEEKED_END_OBJECT) {
+      stackSize--
+      pathNames[stackSize] = null // Free the last path name so that it can be garbage collected!
+      pathIndices[stackSize - 1]++
+      peeked = PEEKED_NONE
+    } else {
+      throw JsonDataException("Expected END_OBJECT but was ${peek()} at path $path")
+    }
+  }
+
+  @Throws(IOException::class)
+  override fun hasNext(): Boolean {
+    var p = peeked
+    if (p == PEEKED_NONE) {
+      p = doPeek()
+    }
+    return p != PEEKED_END_OBJECT && p != PEEKED_END_ARRAY && p != PEEKED_EOF
+  }
+
+  @Throws(IOException::class)
+  override fun peek(): Token {
+    var p = peeked
+    if (p == PEEKED_NONE) {
+      p = doPeek()
+    }
```

Comment:
In a follow-up we should extract this to a `private inline fun`. It appears like 30 times in this file and will eliminate the usage of `var` at all the call-sites.

Reply:
done in 98fdad3